### PR TITLE
Create new front end infrastructure

### PR DESCRIPTION
**Pull request by apmasell**
Here's the base set of infrastructure changes. I will split up the rest of the changes I have to get out more reviewable pieces.

### CODE REVIEW

File: shesmu-server-ui/src/html.ts [hunk 1]
```diff
@@ -0,0 +1,1618 @@
...
+  return buttonCustom(label, title, ["danger"], callback);
+}
+/**
+ * Create an edit button
+ */
+export function buttonEdit(title: string, callback: ClickHandler): UIElement {
+  return buttonSubtle("✎", title, callback);
+}
+/**
+ * Create a button that doesn't have the usual UI style for close and the like
```
Comment (avarsava) [inline shesmu-server-ui/src/html.ts:224]: How is it different other than 'unusual'?

File: shesmu-server-ui/src/html.ts [hunk 2]
```diff
@@ -0,0 +1,1618 @@
...
+    container.removeChild(container.lastChild!);
+  }
+}
+
+/**
+ * A collapsible section
+ * @param title the name of the section
+ * @param inner the contents of the section
+ */
+export function collapse(title: string, ...inner: UIElement[]): UIElement {
```
Comment (avarsava) [inline shesmu-server-ui/src/html.ts:257]: this creates a collapsible section UIElement, but 'collapse()' as a function sounds like an action. Could it be something like 'collapsible()' instead?

File: shesmu-server-ui/src/html.ts [hunk 3]
```diff
@@ -0,0 +1,1618 @@
...
+  };
+  closeButton.addEventListener("click", close);
+  inner.addEventListener("click", (e) => e.stopPropagation());
+  addElements(inner, contents(close));
+}
+/**
+ * Create a drop down list
+ *
+ * This acts like a passive user input that gets read on demand.
+ * @param labelMaker a function to produce a label the user will se for the label
```
Comment (avarsava) [inline shesmu-server-ui/src/html.ts:388]: se?

File: shesmu-server-ui/src/html.ts [hunk 4]
```diff
@@ -0,0 +1,1618 @@
...
+    listElement.appendChild(element);
+  }
+  return container;
+}
+/**
+ * Create a drop down table that can be filtered
+ *
+ * This acts like a passive user input that gets read on demand.
+ * @param model a model to manage this dropdown's state
+ * @param labelMaker a function to produce a label the user will se for the label
```
Comment (avarsava) [inline shesmu-server-ui/src/html.ts:459]: se?

File: shesmu-server-ui/src/html.ts [hunk 5]
```diff
@@ -0,0 +1,1618 @@
...
+/**
+ * A function which will intercept Ctrl-F
+ * @returns true if it wishes to intercept the browser's default behaviour
+ */
+export type FindHandler = (() => boolean) | null;
+
+/**
+ * A mapping type for UI elements that have a multi-pane display
+ */
+export type NamedComponents<T> = {
```
[the remark this answers is not in the document]
Comment (apmasell) [reply shesmu-server-ui/src/html.ts:45]: It's weird TypeScript type magic. I've added a comment. Maybe a different name?

File: shesmu-server-ui/src/html.ts [hunk 6]
```diff
@@ -0,0 +1,1618 @@
...
+  selected?: boolean;
+}
+/**
+ * The contents of a table cell
+ */
+export interface TableCell {
+  click?: ClickHandler;
+  contents: UIElement;
+  header?: boolean;
+  intensity?: number;
```
[the remark this answers is not in the document]
Comment (apmasell) [reply shesmu-server-ui/src/html.ts:94]: Added comments.

File: shesmu-server-ui/src/html.ts [hunk 7]
```diff
@@ -0,0 +1,1618 @@
...
+
+/**
+ * A list that can be updated
+ */
+export interface UpdateableList<T> {
+  add(item: T): void;
+  keepOnly(predicate: (item: T) => boolean): void;
+  replace(items: T[]): void;
+}
+let activeMenu: HTMLElement | null = null;
```
[the remark this answers is not in the document]
Comment (apmasell) [reply shesmu-server-ui/src/html.ts:111]: Unfortunate TypeScript syntax:
- `let activeMenu:` define a variable
- ` HTMLElement | null` whose type is either `HTMLElement` or `null`
- ` = null;` and set it equal to `null`.

File: shesmu-server-ui/src/html.ts [hunk 8]
```diff
@@ -0,0 +1,1618 @@
...
+export function button(
+  label: UIElement,
+  title: string,
+  callback: ClickHandler
+) {
+  return buttonCustom(label, title, [], callback);
+}
+
+/**
+ * Create a button for an unimportant feature
```
[the remark this answers is not in the document]
Comment (apmasell) [reply shesmu-server-ui/src/html.ts:171]: Updated comment.

File: shesmu-server-ui/src/html.ts (hunk 1 above)
Comment (apmasell) [reply shesmu-server-ui/src/html.ts:224]: They're for just icon buttons. I've renamed and added a comment.

File: shesmu-server-ui/src/html.ts (hunk 2 above)
Comment (apmasell) [reply shesmu-server-ui/src/html.ts:257]: Sure. Fixed.

**Pull request merged**